test/test_datapipe.py

```diff
@@ -35,6 +41,63 @@
 
 T_co = TypeVar('T_co', covariant=True)
 
+PORT = 8000
+HOST = 'localhost'
+WEB_TEST_FILE_SIZE = 1024
+WEB_FILE_TEMP_DIR = tempfile.TemporaryDirectory(dir=os.getcwd())
+
+
+def set_up_local_server():
+    # The method sets up a localhost server at current working directory.
+    # The local server starts on a separate thread.
+    # The server is created only once.
+    # It uses a temp folder created above to save temporary files for test(s) of WebIterDataPipe.
+    # By setting the property daemon = True, it would be discarded when the main program exits.
+    #
+    try:
+        Handler = http.server.SimpleHTTPRequestHandler
+        socketserver.TCPServer.allow_reuse_address = True
+
+        server = socketserver.TCPServer(("", PORT), Handler)
+        server_thread = threading.Thread(target=server.serve_forever)
+        server_thread.daemon = True
+        server_thread.start()
+
+        # Wait a bit for the server to come up
+        time.sleep(3)
+        print("Server started ... at {host}:{port}".format(host=HOST, port=PORT))
+    except OSError as e:
+        if e.errno == errno.EADDRINUSE:
+            # localhost:port already in use, no need to initialize again
+            print("There is already an active server serving at {host}:{port}".format(host=HOST, port=PORT))
```

Comment:
Since this method will be invoked for every single method in the TestCase, you will encounter a mount of times of this Error. So, you can use `setUpClass` and `tearDownClass` to run one time per class. And I still think it's better to put them into a separate TestCase, since the server doesn't need to be created if you don't run this test method.
And, please call join for the server thread in the `tearDownClass`.

Reply:
you are right. that setup is called each time a unittest is run. 
fyi it is now moved to a new separate testcase TestIterableDataPipeHttp.